File: lua/birthdayrem.lua
local home = os.getenv('HOME')
local File = io.lines(home .. '/.config/birthdayrem/birthdays')

local curYear = os.date("%Y")
local curMonth = os.date("%m")
local curDay = os.date("%d")

local function parse(file)
	local lines = {}
	local person = {}
	local bYear, bMonth, bDay = "", "", ""
	local dYear, dMonth, dDay = "", "", ""

	for row in file do
		local birthday = string.match(row, "^(%g*),")
		bYear, bMonth, bDay = string.match(birthday, "^(%d*)-(%d*)-(%d*)")

		local name = string.match(row, "%g*,(%W%w+)")

		local deathday = string.match(row, "^%g*,%s%g*,%s(%g*)$")
		if deathday ~= nil then
			dYear, dMonth, dDay = string.match(deathday, "(%d*)-(%d*)-(%d*)")
		else
			dYear, dMonth, dDay = "", "", ""
		end

		person = {
			['name'] = name,
			['bYear'] = bYear,
			['bMonth'] = bMonth,
			['bDay'] = bDay,
			['dYear'] = dYear,
			['dMonth'] = dMonth,
			['dDay'] = dDay,
		}

		table.insert(lines, person)
	end

	return(lines)
end

local function dateCompare(a, b)
	if (a.bMonth < b.bMonth) then
		return true
	elseif (a.bMonth == b.bMonth) then
		if (a.bDay < b.bDay) then
			return true
		elseif (a.bDay == b.bDay) then
			if (a.bYear < b.bYear) then
				return true
			else
				return false
			end
		else
			return false
		end
	end

	return false
end

local function calcAge(sYear, eYear, dYear, after)
	local age = 0

	if (dYear ~= "") then
		if (after) then
			age = (dYear - sYear) - 1
		else
			age = dYear - sYear
		end
	else
		age = eYear - sYear
	end

	return age
end

local function calcNext(sYear, eYear, dYear)
	local next = ""

	if (dYear ~= "") then
		next = "-"
	else
		next = (eYear - sYear) + 1
	end

	return next
end

local function timeSort(people)
	local before = {}
	local after = {}
	local merge = {}

	for _, p in pairs(people) do
		if (p.bMonth < curMonth) then
			table.insert(before, p)
		elseif (p.bMonth == curMonth) then
			if (p.bDay < curDay) then
				table.insert(before, p)
			else
				table.insert(after, p)
			end
		else
			table.insert(after, p)
		end
	end

	for _, p in pairs(after) do
		p['age'] = calcAge(p.bYear, (curYear - 1), p.dYear, true)
		p['next'] = calcNext(p.bYear, (curYear - 1), p.dYear)
		table.insert(merge, p)

	end

	for _, p in pairs(before) do
		p['age'] = calcAge(p.bYear, curYear, p.dYear, false)
		p['next'] = calcNext(p.bYear, curYear, p.dYear)
		table.insert(merge, p)
	end

	return merge
end

local function output(people)
	print(string.format(
			"%4s-%2s-%2s %10s %8s %8s",
			"yyyy", "mm", "dd", "NAME", "AGE-NOW", "NEXT"
		)
	)

	print("---------------------------------------")

	for _, p in pairs(people) do
		print(string.format(
			"%4s-%2s-%2s %10s %8s %8s",
			p['bYear'], p['bMonth'], p['bDay'], p['name'], p['age'], p['next']
			)
		)
	end
end

local people = parse(File)
table.sort(people, dateCompare)
people = timeSort(people)
output(people)

File: node/birthdayrem.js
#!/usr/bin/env node

const util = require('util')

const fs = require('fs')
const homedir = require('os').homedir()
const birthdayFile = homedir + '/.config/birthdayrem/birthdays'
const people = []
const now = []


fs.readFile(birthdayFile, 'utf8', (err, data) => {
	if (err) {
		console.error(err)
		return
	}

	const rows = data.toString().split('\n');

	for (i in rows) {
		if (rows[i] != '') {
			people[i] = parseRows(rows[i])
		}
	}

	sortByDate()
	setNow()
	const p = sortHasHadBirthday()
	const pWithAges = calcAges(p)
	output(pWithAges)
});


function calcAges(peeps) {
	const ret = []

	for (i in peeps) {
		const p = peeps[i]

		ret[i] = p

		if (p['deathday'].length == 0) {
			if (p['hadBirthday']) {
				ret[i]['age'] = now['year'] - p['bYear']
			} else {
				ret[i]['age'] = (now['year'] - p['bYear']) - 1
			}

			ret[i]['next'] = ret[i]['age'] + 1
		} else {
			if (p['hadBirthday']) {
				ret[i]['age'] = p['dYear'] - p['bYear']
			} else {
				ret[i]['age'] = (p['dYear'] - p['bYear']) - 1
			}

			ret[i]['next'] = '-'
		}
	}

	return ret
}


function setNow() {
	const date = new Date();
    now['year'] = date.getUTCFullYear()
	now['month'] = date.getUTCMonth() + 1
	now['day'] = date.getUTCDate()
}


function sortHasHadBirthday() {
	const before = []
	const after = []
	const ret = []
	var countBefore = 0
	var countAfter = 0
	var countRet = 0

	for (i in people) {
		const p = people[i]
		p['hadBirthday'] = false

		if (p['bMonth'] < now['month']) {
			p['hadBirthday'] = true
			before[countBefore++] = p
		} else if (p['bMonth'] == now['month']) {
			if (p['bDay'] < now['day']) {
				p['hadBirthday'] = true
				before[countBefore++] = p
			} else {
				after[countAfter++] = p
			}
		} else {
			after[countAfter++] = p
		}
	}

	for (i = 0; i < countAfter; i++) {
		ret[countRet++] = after[i]
	}

	for (i = 0; i < countBefore; i++) {
		ret[countRet++] = before[i]
	}

	return ret
}


function sortByDate() {
	people.sort(function (a, b) {
		if (a['bMonth'] < b['bMonth']) {
			return -1
		} else if (a['bMonth'] == b['bMonth']) {
			if (a['bDay'] < b['bDay']) {
				return -1
			} else if (a['bDay'] == b['bDay']) {
				if (a['bYear'] < b['bYear']) {
					return -1
				} else {
					return 1
				}
			} else {
				return 1
			}
		} else {
			return 1
		}
	})
}


function parseRows(row) {
	const ret = [];
	const rowArray = row.split(',')

	ret['birthday'] = rowArray[0]

	const bArray = ret['birthday'].split('-')

	ret['bYear'] = bArray[0]
	ret['bMonth'] = bArray[1]
	ret['bDay'] = bArray[2]

	ret['name'] = rowArray[1].trim()

	if (rowArray[2]) {
		ret['deathday'] = rowArray[2]

		const dArray = ret['deathday'].split('-')

		ret['dYear'] = dArray[0]
		ret['dMonth'] = dArray[1]
		ret['dDay'] = dArray[2]
	} else {
		ret['deathday'] = ''
		ret['dYear'] = ''
		ret['dMonth'] = ''
		ret['dDay'] = ''
	}

	return ret
}


function output(ppl) {
	console.log('yyyy-mm-dd       NAME  AGE-NOW     NEXT')
	console.log('---------------------------------------')

	for (i in ppl) {
		const p = ppl[i]
		console.log(
			util.format(
				'%s %s %s %s', 
				p['birthday'], 
				p['name'].padStart(10),
				p['age'].toString().padStart(8),
				p['next'].toString().padStart(8),
			)
		)
	}
}
